callgraph/core.py

```diff
@@ -187,6 +194,34 @@ def _AnnotateNode(self, node):
                     node.AddConnection(target, command, line_number)
                     print(u"Line {} has a goto towards: <{}>. Current block: {}".format(line_number, target, node.name), file=self.log_file)
 
+                # cgreen - external calls enhancements
+                if command == "external_call":
+                    # cgreen -
+                    # check if the external call is to another batch file or to some other program
+                    # rudimentary check of the file extension
+                    cmdext = target.rsplit('.',1)[1] 
+                    if cmdext=="bat" or cmdext=='cmd':
+                        node.AddExternalConnection(target, command, line_number)                  
+                        print(u"!! Line {} has a external script call towards: <{}>. Current block: {}".format(line_number, target, node.name), file=self.log_file)
+                        # cgreen -
+                        # need a better way to determine if the call to the new cmd has a path specified
+                        #  or does it assume the cmd is in the same folder as the inital script
+                        #
+                        # if follow flag is enabled then go parse that file too..
+                        if 1==1:
+                            print(u"Follow calls would go here..")
+                            cmdfile = open(target, 'r')
+                            CallGraph.Build(cmdfile, log_file=sys.stderr)
```

Comment:
The `CallGraph.Build` method will return an instance of `CallGraph`. I think you'll need to store those graphs in a dedicated data structure in the `CallGraph` object itself.

This will allow you to refer to those nested call graphs when representing the call graph.

As a first step, I suggest to maybe use a dictionary mapping the absolute file name to the generated call graph instance.

Reply:
As external scripts are being parsed, the new dict in CallGraph is being set, and I believe this is being nested representing the levels of sub-calls .. and being rendered through PrintDot making a recursive call to the new method to output the dot contents - however it appears the output png, while now displaying the nodes, they appear to be rendered at the top node. example attached 
md calls a third level loc.cmd
![example1](https://user-images.githubusercontent.com/972358/94373923-b1beb980-00d6-11eb-9e1b-3fe39c03a8bb.png)

Reply:
I think this happens because in `render.py` you are creating the dot representation of the call graphs (lines 31-33), but I don't see any place where you are creating a connection between the node that does the call to the sub-graph corresponding to the given call.

You might need to walk the `connections` attribute of the `CallGraph` object and create extra lines that way.

Adding another minor node that we may want to tackle in a later iteration: what happens if the more than one script has a given label? We may need to do something clever in the dot representation. We should worry about this later, writing it down so I don't forget :)

Reply:
This is looking a little better, the correct nodes are being connected - not sure if we want to keep the change to the '__begin__' nodes where it is now concatenated with file name. I also tried to look into rendering the call to the external file as a rectangle however I wasn't able to make progress there in the time available.

Also noticed in the example png, the node with the external call to 'loc.cmd' is being treated differently, I wasn't able to identify as of yet why those characters are being stripped.

![example1](https://user-images.githubusercontent.com/972358/94389219-42b88380-011d-11eb-8041-cab33e0c53e5.png)

Reply:
> This is looking a little better, the correct nodes are being connected - not sure if we want to keep the change to the 'begin' nodes where it is now concatenated with file name

What you could do is to encode the file name in the node ID only, and using the `label` graphviz attribute to render a shorter text. For example, instead of:

```
"__begin__c:\dev\cmd-call-graph\examples\x.cmd" [label=<<b>__begin__c:\dev\cmd-call-graph\examples\x.cmd</b><br/>(line 1)>]
```

we could print in the output:

```
"__begin__c:\dev\cmd-call-graph\examples\x.cmd" [label=<<b>__begin__</b><br/>(line 1)>]
```

This will preserve the correct edges while improving the appearance of the output.

> I also tried to look into rendering the call to the external file as a rectangle however I wasn't able to make progress there in the time available.

Take your time! You are making great progress. 👍 

> Also noticed in the example png, the node with the external call to 'loc.cmd' is being treated differently, I wasn't able to identify as of yet why those characters are being stripped.

We might need to quote the node names and labels, as now they can contain paths which might contain escape sequences, see for example https://stackoverflow.com/a/18445123.

Reply:
Improved the begin labels as suggested to improve the output - also using a new example of several nested external calls (a calls b> calls c> calls d> calls e). 

I think the symptom of that node that appears to need to be escaped in the previous message is also a side effect of something happening in either the annotations or the removing of connections, possibly related to those recognized as terminating or exit node.
![a](https://user-images.githubusercontent.com/972358/94485465-f6685480-01ab-11eb-83d6-c62fd9f50c42.png)
 

```
Follow flag enabled
*** Calling _ParseSource:<_io.TextIOWrapper name='C:\\dev\\cmd-call-graph/examples/a.cmd' mode='r' encoding='cp1252'>

*** Calling Build on input file:<_io.TextIOWrapper name='c:\\dev\\cmd-call-graph\\examples\\b.cmd' mode='r' encoding='cp1252'>
*** Calling _ParseSource:<_io.TextIOWrapper name='c:\\dev\\cmd-call-graph\\examples\\b.cmd' mode='r' encoding='cp1252'>
Line 5 defines a new block: <bwork>
Annotating node __begin__c:\dev\cmd-call-graph\examples\b.cmd (line 1)
Line 3 has a goto towards: <bwork>. Current block: __begin__c:\dev\cmd-call-graph\examples\b.cmd
Annotating node c:\dev\cmd-call-graph\examples\b.cmd (line -1)
Annotating node eof (line -1)
Annotating node bwork (line 5)
!! Line 6 has a external script call towards: <c:\dev\cmd-call-graph\examples\c.cmd>. Current block: bwork

*** Calling Build on input file:<_io.TextIOWrapper name='c:\\dev\\cmd-call-graph\\examples\\c.cmd' mode='r' encoding='cp1252'>
*** Calling _ParseSource:<_io.TextIOWrapper name='c:\\dev\\cmd-call-graph\\examples\\c.cmd' mode='r' encoding='cp1252'>
Line 5 defines a new block: <cwork>
Annotating node __begin__c:\dev\cmd-call-graph\examples\c.cmd (line 1)
Line 3 has a goto towards: <cwork>. Current block: __begin__c:\dev\cmd-call-graph\examples\c.cmd
Annotating node c:\dev\cmd-call-graph\examples\c.cmd (line -1)
Annotating node eof (line -1)
Annotating node cwork (line 5)
!! Line 6 has a external script call towards: <c:\dev\cmd-call-graph\examples\d.cmd>. Current block: cwork

*** Calling Build on input file:<_io.TextIOWrapper name='c:\\dev\\cmd-call-graph\\examples\\d.cmd' mode='r' encoding='cp1252'>
*** Calling _ParseSource:<_io.TextIOWrapper name='c:\\dev\\cmd-call-graph\\examples\\d.cmd' mode='r' encoding='cp1252'>
Line 5 defines a new block: <dwork>
Annotating node __begin__c:\dev\cmd-call-graph\examples\d.cmd (line 1)
Line 3 has a goto towards: <dwork>. Current block: __begin__c:\dev\cmd-call-graph\examples\d.cmd
Annotating node c:\dev\cmd-call-graph\examples\d.cmd (line -1)
Annotating node eof (line -1)
Annotating node dwork (line 5)
!! Line 6 has a external script call towards: <c:\dev\cmd-call-graph\examples\e.cmd>. Current block: dwork

*** Calling Build on input file:<_io.TextIOWrapper name='c:\\dev\\cmd-call-graph\\examples\\e.cmd' mode='r' encoding='cp1252'>
*** Calling _ParseSource:<_io.TextIOWrapper name='c:\\dev\\cmd-call-graph\\examples\\e.cmd' mode='r' encoding='cp1252'>
Line 6 defines a new block: <eworkdone>
Annotating node __begin__c:\dev\cmd-call-graph\examples\e.cmd (line 1)
Line 3 has a goto towards: <eworkdone>. Current block: __begin__c:\dev\cmd-call-graph\examples\e.cmd
Annotating node c:\dev\cmd-call-graph\examples\e.cmd (line -1)
Annotating node eof (line -1)
Annotating node eworkdone (line 6)

Removing the eof node, since there are no call/nested connections to it and it's not a real node
Removing 0 eof connections in node __begin__c:\dev\cmd-call-graph\examples\e.cmd
Removing 0 eof connections in node c:\dev\cmd-call-graph\examples\e.cmd
Removing 0 eof connections in node eworkdone
eworkdone is the last node, marking it as exit node.

Removing the eof node, since there are no call/nested connections to it and it's not a real node
Removing 0 eof connections in node __begin__c:\dev\cmd-call-graph\examples\d.cmd
Removing 0 eof connections in node c:\dev\cmd-call-graph\examples\d.cmd
Removing 0 eof connections in node dwork
dwork is the last node, marking it as exit node.

Removing the eof node, since there are no call/nested connections to it and it's not a real node
Removing 0 eof connections in node __begin__c:\dev\cmd-call-graph\examples\c.cmd
Removing 0 eof connections in node c:\dev\cmd-call-graph\examples\c.cmd
Removing 0 eof connections in node cwork
cwork is the last node, marking it as exit node.

Removing the eof node, since there are no call/nested connections to it and it's not a real node
Removing 0 eof connections in node __begin__c:\dev\cmd-call-graph\examples\b.cmd
Removing 0 eof connections in node c:\dev\cmd-call-graph\examples\b.cmd
Removing 0 eof connections in node bwork
bwork is the last node, marking it as exit node.
```

Reply:
Added recursive call while rendering to get all the nested dictionaries to build entire graph
<img width="156" alt="fixing_problems" src="https://user-images.githubusercontent.com/972358/94508568-347f6b80-01e0-11eb-9ed3-e05166d66cc6.png">